### PR TITLE
Handles GitHub package authentication errors

### PR DESCRIPTION
**Pull request by sachin-sandhu**
### What are you trying to accomplish?

Ecosystem : NPM and YARN
Exception Count : 3.2k (Since April 2024)

Preface: Resolving issues related with Dependabot::SharedHelpers::HelperSubprocessFailed issues

Issue: Dependabot will throw uknown_error in case a github package registry (hosted at https://npm.pkg.github.com) is inaccessible due to missing or invalid authentication information. Right now, the error is not captured and is thrown into Dependabot::SharedHelpers::HelperSubprocessFailed unknown_error . Following is variation of errors that is returned from native helper:

`Missing or invalid authenticaton token while accessing github package https://npm.pkg.github.com/@Codertocat%2fhello-world-npm`

`401 Unauthorized - GET https://npm.pkg.github.com/@Codertocat%2fhello-world-npm - unauthenticated: User cannot be authenticated with the token provided.`

Error log:

```
+-------------------------------+
| Dependencies failed to update |
+---------------+---------------+
| js           | unknown_error |
+---------------+---------------+
```

Fix: To capture the response returned from native helper and raise git_token_auth_error exception instead.

```
+---------------------------------------------------+
|           Dependencies failed to update           |
+-----------+---------------------------------------+
| io | git_token_auth_error                         |
+-----------+---------------------------------------+

```

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

File: common/lib/dependabot/errors.rb [hunk 1]
```diff
@@ -531,6 +538,20 @@ def initialize(source)
     end
   end
 
+  class GitAuthToken < DependabotError
```
Comment (abdulapopoola) [inline common/lib/dependabot/errors.rb:541]: Can we have a more descriptive error name? E.g. `InvalidGitAuthToken` or something like that

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb [hunk 2]
```diff
@@ -74,6 +74,10 @@ def updated_lockfile_reponse(response)
         SOCKET_HANG_UP = /request to (?<url>.*) failed, reason: socket hang up/
         UNABLE_TO_AUTH_NPMRC = /Unable to authenticate, need: Basic, Bearer/
         UNABLE_TO_AUTH_REGISTRY = /Unable to authenticate, need: *.*(Basic|BASIC) *.*realm="(?<url>.*)"/
+        MISSING_AUTH_TOKEN = /401 Unauthorized - GET (?<url>.*) - authentication token not provided/
+        INVALID_AUTH_TOKEN =
+          /401 Unauthorized - GET (?<url>.*) - unauthenticated: User cannot be authenticated with the token provided./
+        GIT_PACKAGE = "https://npm.pkg.github.com"
```
Comment (abdulapopoola) [inline npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb:80]: is this only applicable to npm?

Comment (abdulapopoola) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb:80]: If so, should we rename to `NPM PACKAGE REGISTRY` or something like that?

File: common/lib/dependabot/errors.rb (hunk 1 above)
Comment (sachin-sandhu) [reply common/lib/dependabot/errors.rb:541]: makes sense, ill change it

File: common/lib/dependabot/errors.rb [hunk 3]
```diff
@@ -230,6 +230,11 @@ def self.updater_error_details(error)
           message: error.message
         }
       }
+    when Dependabot::GitAuthToken
```
Comment (abdulapopoola) [inline common/lib/dependabot/errors.rb:233]: seems this is a new error, thus it'll need to handled E2E

File: npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb (hunk 2 above)
Comment (sachin-sandhu) [reply npm_and_yarn/lib/dependabot/npm_and_yarn/file_updater/npm_lockfile_updater.rb]: fixed

File: common/lib/dependabot/errors.rb (hunk 3 above)
Comment (sachin-sandhu) [reply common/lib/dependabot/errors.rb]: ill raise API PR for this

**Review by abdulapopoola — APPROVED** [review]
Please address the codeql issues

**Pull request merged**